File: fastFramework/src/main/java/com/tesla/framework/ui/activity/BaseActivity.java
package com.tesla.framework.ui.activity;

import android.app.Activity;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.orhanobut.logger.Logger;
import com.tesla.framework.R;
import com.tesla.framework.component.network.NetworkStateManager;
import com.tesla.framework.ui.fragment.BaseStatusFragmentNew;
import com.tesla.framework.ui.widget.CustomToolbar;

import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelStoreOwner;


/**
 * Created by JerryLiu on 17/04/08.
 */

public abstract class BaseActivity extends AppCompatActivity implements CustomToolbar.OnToolbarDoubleClickListener {

    static final String TAG = "Activity-Base";

    private BaseActivityHelper mHelper;

    // 当有Fragment Attach到这个Activity的时候，就会保存
    private Map<String, WeakReference<BaseStatusFragmentNew>> fragmentRefs;


    private ViewModelProvider mActivityProvider;
    private ViewModelProvider mApplicationProvider;


    /**
     * 判断当前Activity是否在前台。
     */
    protected boolean isActive = false;

    /**
     * 当前Activity的实例。
     */
    protected Activity activity = null;

    private WeakReference<Activity> weakRefActivity = null;


    private View rootView;


    Toolbar mToolbar;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        fragmentRefs = new HashMap<>();
        super.onCreate(savedInstanceState);
        activity = this;
        weakRefActivity = new WeakReference(this);

        if (getLayoutId() > 0) {
            setContentView(getLayoutId());
        }

        getLifecycle().addObserver(NetworkStateManager.getInstance());
//        BarUtils.setStatusBarColor(this, Color.TRANSPARENT);
//        BarUtils.setStatusBarLightMode(this, true);
        bindView();
        layoutInit(savedInstanceState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        isActive = true;
    }

    @Override
    protected void onPause() {
        super.onPause();
        isActive = false;
    }

    protected void bindView(){
        mToolbar = findViewById(R.id.toolbar);
    }


    public void showContent(Class<? extends BaseStatusFragmentNew> target) {
        showContent(target, null);
    }

    public void showContent(Class<? extends BaseStatusFragmentNew> target, Bundle bundle) {
        try {
            BaseStatusFragmentNew fragment = target.newInstance();
            if (bundle != null) {
                fragment.setArguments(bundle);
            }
            FragmentManager fm = getSupportFragmentManager();
            FragmentTransaction fragmentTransaction = fm.beginTransaction();
            fragmentTransaction.add(android.R.id.content, fragment);
            fragmentTransaction.addToBackStack("");
            fragmentTransaction.commit();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
    }


    /**
     * 指定Fragment的LayoutID
     *
     * @return
     */
    public abstract int getLayoutId();


    /**
     * 子类重写这个方法，初始化视图
     */
    public void layoutInit(Bundle savedInstanceState) {

    }


    public Toolbar getToolbar() {
        return mToolbar;
    }

    public View getRootView() {
        return rootView;
    }

    @Override
    public void setContentView(int layoutResID) {
        setContentView(View.inflate(this, layoutResID, null));
    }

    @Override
    public void setContentView(View view, ViewGroup.LayoutParams params) {
        super.setContentView(view, params);
        rootView = view;
        if (mToolbar != null)
            setSupportActionBar(mToolbar);
    }

    @Override
    public void setContentView(View view) {
        super.setContentView(view, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        rootView = view;
        mToolbar = findViewById(R.id.toolbar);
        if (mToolbar != null)
            setSupportActionBar(mToolbar);
    }


    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);



    }

    public void addFragment(String tag, BaseStatusFragmentNew fragment) {
        fragmentRefs.put(tag, new WeakReference<>(fragment));
    }

    public void removeFragment(String tag) {
        fragmentRefs.remove(tag);
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (mHelper != null) {
            boolean handle = mHelper.onOptionsItemSelected(item);
            if (handle){
                return true;
            }
        }

        if (item.getItemId() == android.R.id.home) {
            if (onHomeClick())
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    protected boolean onHomeClick() {
        if (getActivityHelper() != null) {
            boolean handle = getActivityHelper().onHomeClick();
            if (handle)
                return true;
        }

        Set<String> keys = fragmentRefs.keySet();
        for (String key : keys) {
            WeakReference<BaseStatusFragmentNew> fragmentRef = fragmentRefs.get(key);
            BaseStatusFragmentNew fragment = fragmentRef.get();
            if (fragment != null && fragment.onHomeClick())
                return true;
        }

        return onBackClick();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (mHelper != null) {
            boolean handle = mHelper.onKeyDown(keyCode, event);
            if (handle)
                return true;
        }

        if (keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount() == 0) {
            if (onBackClick())
                return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    public boolean onBackClick() {
        if (getActivityHelper() != null) {
            boolean handle = getActivityHelper().onBackClick();
            if (handle)
                return true;
        }

        Set<String> keys = fragmentRefs.keySet();
        for (String key : keys) {
            WeakReference<BaseStatusFragmentNew> fragmentRef = fragmentRefs.get(key);
            BaseStatusFragmentNew fragment = fragmentRef.get();
            if (fragment != null && fragment.onBackClick())
                return true;
        }

        finish();

        return true;
    }


    @Override
    public boolean OnToolbarDoubleClick() {
        Set<String> keys = fragmentRefs.keySet();
        for (String key : keys) {
            WeakReference<BaseStatusFragmentNew> fragmentRef = fragmentRefs.get(key);
            BaseStatusFragmentNew fragment = fragmentRef.get();
            if (fragment != null && fragment instanceof CustomToolbar.OnToolbarDoubleClickListener) {
                if (((CustomToolbar.OnToolbarDoubleClickListener) fragment).OnToolbarDoubleClick())
                    return true;
            }
        }

        return false;
    }


    public BaseActivityHelper getActivityHelper() {
        return mHelper;
    }


    protected void setToolbarTitle(String msg) {
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(msg);

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        activity = null;
    }


    protected <T extends ViewModel> T getActivityViewModel(@NonNull Class<T> modelClass) {
        if (mActivityProvider == null) {
            mActivityProvider = new ViewModelProvider(this);
        }
        return mActivityProvider.get(modelClass);
    }

    protected <T extends ViewModel> T getApplicationScopeViewModel(@NonNull Class<T> modelClass) {
        if (mApplicationProvider == null) {
            mApplicationProvider = new ViewModelProvider((ViewModelStoreOwner) this.getApplicationContext());
        }
        return mApplicationProvider.get(modelClass);
    }


    /**
     * https://t.zsxq.com/uRBEaEE
     * @param overrideConfiguration
     */
    @Override
    public void applyOverrideConfiguration(Configuration overrideConfiguration) {
        Logger.d("applyOverrideConfiguration: %s ");
        if (overrideConfiguration != null){
            //
           /* if (LocalMananger.locale != null){
                overrideConfiguration.setLocale(Loca);
            }*/
        }
        super.applyOverrideConfiguration(overrideConfiguration);
    }
}

File: fastFramework/src/main/java/com/tesla/framework/ui/activity/BaseActivity.kt
package com.tesla.framework.ui.activity

import android.app.Activity
import android.content.Intent
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.Color
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelStoreOwner
import androidx.preference.PreferenceManager
import com.blankj.utilcode.util.ToastUtils
import com.gyf.immersionbar.ktx.immersionBar
import com.tesla.framework.R
import com.tesla.framework.common.util.AndroidBugFixUtils
import com.tesla.framework.component.hook.HookResources
import com.tesla.framework.component.logger.Logger
import com.tesla.framework.component.network.AutoRegisterNetListener
import com.tesla.framework.component.network.NetworkTypeEnum
import com.tesla.framework.performance.takt.Takt
import com.tesla.framework.ui.fragment.BaseFragment
import com.tesla.framework.ui.fragment.base.IDynamicView
import com.tesla.framework.ui.widget.CustomToolbar.OnToolbarDoubleClickListener
import com.zwb.lib_base.utils.network.NetworkStateChangeListener
import java.lang.ref.WeakReference

/**
 * Created by Jerry on 2022/11/26.
 */
open abstract class BaseActivity: AppCompatActivity(), OnToolbarDoubleClickListener, IDynamicView {
    private  var mDelegate: BaseActivityDelegate ?= null

    // 当有Fragment Attach到这个Activity的时候，就会保存
    private var fragmentRefs: MutableMap<String, WeakReference<BaseFragment>>? = null
    private var mActivityProvider: ViewModelProvider? = null
    private var mApplicationProvider: ViewModelProvider? = null



    /**
     * 当前Activity的实例。
     */
    protected var activity: Activity? = null
    private var toolbar: Toolbar? = null
    protected lateinit var root: View


    override fun onCreate(savedInstanceState: Bundle?) {
        fragmentRefs = HashMap()
        super.onCreate(savedInstanceState)
        activity = this


        initViewModel()
        initNetworkListener()

        inflateView().also {
            root = it
        }

        initView(root)
        layoutInit(savedInstanceState)

    }


    private fun getHelperDelegate():BaseActivityDelegate{
        if (mDelegate == null){
            mDelegate =
                BaseActivityDelegate(this, this)
        }
        return mDelegate!!
    }

    override fun onStart() {
        super.onStart()
        getHelperDelegate().onStart(this)
        Takt.play()
    }

    override fun onResume() {
        super.onResume()
        getHelperDelegate().onResume(this)
    }

    override fun onPause() {
        super.onPause()
        getHelperDelegate().onPause(this)
    }

    override fun onStop() {
        super.onStop()
        getHelperDelegate().onStop(this)
//        Takt.finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        getHelperDelegate().onDestroy(this)
        activity = null
        AndroidBugFixUtils.fixSoftInputLeaks(this)
    }


    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)

    }
    open abstract fun inflateView():View





    /**
     * 初始化网络状态监听
     * @return Unit
     */
    private fun initNetworkListener() {
        lifecycle.addObserver(AutoRegisterNetListener(object :NetworkStateChangeListener{
            override fun networkTypeChange(type: NetworkTypeEnum) {
                println("BaseVBActivity networkTypeChange type:$type")
                ToastUtils.showShort("网络状况: $type")
            }


            override fun networkConnectChange(isConnected: Boolean) {
                println("BaseVBActivity networkConnectChange isConnected:$isConnected")
                ToastUtils.showShort("网络状况: $isConnected")
            }

        }))

    }


    /**
     * 状态栏导航栏初始化
     */
    private fun initSystemBar() {
        immersionBar {
            transparentStatusBar()
            statusBarDarkFont(true)
            navigationBarColor(R.color.white)
            navigationBarDarkIcon(true)
        }
    }

    open fun setUpActionBar() {
        toolbar = findViewById(R.id.toolbar)
        if (toolbar != null) {
            setSupportActionBar(toolbar)

        }
        supportActionBar?.apply {
            //使左上角图标是否显示，如果设成false，则没有程序图标，仅仅就个标题，否则，显示应用程序图标，对应id为Android.R.id.home，对应ActionBar.DISPLAY_SHOW_HOME
            setDisplayShowHomeEnabled(true)
            //给左上角图标的左边加上一个返回的图标 。对应ActionBar.DISPLAY_HOME_AS_UP
            setDisplayHomeAsUpEnabled(true)

        }
    }

    open fun initView(rootView: View?) {
        window.statusBarColor = Color.TRANSPARENT // don't draw a status bar, the DrawerLayout and the MaterialDrawerLayout have their own

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        preferences.getString("statusTextSize", "medium")?.let {
            theme.applyStyle(textStyle(it), true)
        }

    }

    private  fun textStyle(name: String): Int {
        val style: Int
        when (name) {
            "smallest" -> style = R.style.TextSizeSmallest
            "small" -> style = R.style.TextSizeSmall
            "medium" -> style = R.style.TextSizeMedium
            "large" -> style = R.style.TextSizeLarge
            "largest" -> style = R.style.TextSizeLargest
            else -> style = R.style.TextSizeMedium
        }
        return style
    }


    open fun startActivityWithSlideInAnimation(intent: Intent?) {
        super.startActivity(intent)
        overridePendingTransition(R.anim.slide_in_left, R.anim.slide_out_right)
    }

    @JvmOverloads
    fun showContent(target: Class<out BaseFragment>, bundle: Bundle? = null) {
        try {
            val fragment = target.newInstance()
            if (bundle != null) {
                fragment.arguments = bundle
            }
            val fm = supportFragmentManager
            val fragmentTransaction = fm.beginTransaction()
            fragmentTransaction.add(android.R.id.content, fragment)
            fragmentTransaction.addToBackStack("")
            fragmentTransaction.commit()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }

    /**
     * 子类重写这个方法，初始化视图
     */
    open fun layoutInit(savedInstanceState: Bundle?) {}
    fun addFragment(tag: String, fragment: BaseFragment) {
        fragmentRefs!![tag] = WeakReference(fragment)
    }

    fun removeFragment(tag: String) {
        fragmentRefs!!.remove(tag)
    }

    /**
     * 暂时不实现这个方法，避免使用 drawalayout 时无法打开抽屉，满足了
     * item.itemId == android.R.id.home 的条件无法打开抽屉
     */
    /* override fun onOptionsItemSelected(item: MenuItem): Boolean {
         if (activityHelper != null) {
             val handle = activityHelper.onOptionsItemSelected(item)
             if (handle) {
                 return true
             }
         }
         if (item.itemId == android.R.id.home) {
             if (onHomeClick()) return true
         }
         return super.onOptionsItemSelected(item)
     }*/

    protected fun onHomeClick(): Boolean {
        if (mDelegate != null) {
            val handle = getHelperDelegate().onHomeClick()
            if (handle) return true
        }
        val keys: Set<String> = fragmentRefs!!.keys
        for (key in keys) {
            val fragmentRef = fragmentRefs!![key]!!
            val fragment = fragmentRef.get()
            if (fragment != null && fragment.onHomeClick()) return true
        }
        return onBackClick()
    }

    /*override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (activityHelper != null) {
            val handle = activityHelper.onKeyDown(keyCode, event)
            if (handle) return true
        }
        if (keyCode == KeyEvent.KEYCODE_BACK && event.repeatCount == 0) {
            if (onBackClick()) return true
        }
        return super.onKeyDown(keyCode, event)
    }*/

    open fun onBackClick(): Boolean {
        if (mDelegate != null) {
            val handle = getHelperDelegate().onBackClick()
            if (handle) return true
        }
        val keys: Set<String> = fragmentRefs!!.keys
        for (key in keys) {
            val fragmentRef = fragmentRefs!![key]!!
            val fragment = fragmentRef.get()
            if (fragment != null && fragment.onBackClick()) return true
        }
        finish()
        return true
    }

    override fun OnToolbarDoubleClick(): Boolean {
        val keys: Set<String> = fragmentRefs!!.keys
        for (key in keys) {
            val fragmentRef = fragmentRefs!![key]!!
            val fragment = fragmentRef.get()
            if (fragment != null && fragment is OnToolbarDoubleClickListener) {
                if ((fragment as OnToolbarDoubleClickListener).OnToolbarDoubleClick()) return true
            }
        }
        return false
    }

    fun setToolbarTitle(msg: String?) {
        val supportActionBar = supportActionBar
        if (supportActionBar != null) {
            supportActionBar.title = msg
        }
    }


    protected fun <T : ViewModel> getActivityViewModel(modelClass: Class<T>): T {
        if (mActivityProvider == null) {
            mActivityProvider = ViewModelProvider(this)
        }
        return mActivityProvider!![modelClass]
    }

    protected fun <T : ViewModel> getApplicationScopeViewModel(modelClass: Class<T>): T {
        if (mApplicationProvider == null) {
            mApplicationProvider =
                ViewModelProvider((this.applicationContext as ViewModelStoreOwner))
        }
        return mApplicationProvider!![modelClass]
    }

    /**
     * https://t.zsxq.com/uRBEaEE
     * @param overrideConfiguration
     */
    override fun applyOverrideConfiguration(overrideConfiguration: Configuration) {
        Logger.d("applyOverrideConfiguration: %s ")
        if (overrideConfiguration != null) {
            //
            /* if (LocalMananger.locale != null){
                overrideConfiguration.setLocale(Loca);
            }*/
        }
        super.applyOverrideConfiguration(overrideConfiguration)
    }

    protected open fun initViewModel() {}

    companion object {
        const val TAG = "Activity-Base"

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
//            onBackPressedDispatcher.onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }




    fun showToast(msg: String?) {
        msg?.let {
            ToastUtils.showShort(msg)
        }
    }

    fun showToast(@StringRes msgId: Int) {
        showToast(getString(msgId))
    }

//    override fun getBaseContext(): Context {
//        val originContext =  super.getBaseContext()
//        return HookContext(originContext)
//    }



    private var mHookResources: HookResources?= null


    override fun getResources(): Resources {
        val originRes = super.getResources()
        if (mHookResources == null){
            mHookResources = HookResources(originRes)
        }
        val result = mHookResources!!
        if (result.configuration != originRes.configuration || result.displayMetrics != originRes.displayMetrics){
            result.updateConfiguration(originRes.configuration,originRes.displayMetrics)
        }
        return result
    }


    fun isActive():Boolean = getHelperDelegate().isActive


}